rafs/src/metadata/cached_v5.rs

```diff
@@ -355,27 +355,36 @@ impl CachedInodeV5 {
 }
 
 impl RafsInode for CachedInodeV5 {
-    #[allow(clippy::collapsible_if)]
+    // Somehow we got invalid `inode_count` from superblock.
     fn validate(&self, _inode_count: u64, chunk_size: u64) -> Result<()> {
-        if self.i_ino == 0 || self.i_name.len() > RAFS_MAX_NAME || self.i_nlink == 0 {
+        if self.i_ino == 0
+            // || self.i_ino > inode_count
+            || self.i_nlink == 0
+            || (self.i_ino != RAFS_V5_ROOT_INODE && self.i_parent == 0)
+            || self.i_name.len() > RAFS_MAX_NAME
+            || self.i_name.is_empty()
+        {
             return Err(einval!("invalid inode"));
         }
+        if !self.is_hardlink() && self.i_parent >= self.i_ino {
+            return Err(einval!("invalid parent inode"));
+        }
         if self.is_reg() {
             let chunks = (self.i_size + chunk_size - 1) / chunk_size;
             if !self.has_hole() && chunks != self.i_data.len() as u64 {
                 return Err(einval!("invalid chunk count"));
             }
+            let blocks = (self.i_size + 511) / 512;
+            // Old stargz builder generates inode with 0 blocks
+            if blocks != self.i_blocks && self.i_blocks != 0 {
```

Comment:
It seems we calculate the blocks by:

```
self.inode.set_blocks(div_round_up(
  self.inode.size() + self.xattrs.aligned_size_v5() as u64,
  512,
));
```

Reply:
Seems that code path only works for directory builder, stargz builder has its own implementation.

Reply:
We should fix stargz builder.